waku/v2/utils/credentials.nim

```diff
@@ -0,0 +1,397 @@
+                        version: string,
+                        separator: string = "\n"): KeystoreResult[void] =
+
+  let keystore = AppKeystore(application: application,
+                             appIdentifier: appIdentifier,
+                             credentials: @[],
+                             version: version)
+
+  var jsonKeystore: string
+  jsonKeystore.toUgly(%keystore)
+
+  var f: File
+  if not f.open(path, fmWrite):
+    return err(OsError)
+
+  try:
+    # To avoid other users/attackers to be able to read keyfiles, we make the file readable/writable only by the running user
+    setFilePermissions(path, {fpUserWrite, fpUserRead})
+    f.write(jsonKeystore)
+    # We separate keystores with separator
+    f.write(separator)
+    ok()
+  except CatchableError:
+    err(OsError)
+  finally:
+    f.close()
+
+# This proc load a keystore based on the application, appIdentifier and version filters. 
+# If none is found, it automatically creates an empty keystore for the passed parameters
+proc loadAppKeystore*(path: string,
+                      application: string,
+                      appIdentifier: string,
+                      version: string,
+                      separator: string = "\n"): KeystoreResult[JsonNode] =
+
+  ## Load and decode JSON keystore from pathname
+  var data: JsonNode
+  var matchingAppKeystore: JsonNode
+
+  # If no keystore exists at path we create a new empty one with passed keystore parameters
+  if fileExists(path) == false:
+    let newKeystore = createAppKeystore(path, application, appIdentifier, version, separator)
+    if newKeystore.isErr():
+        return err(CreateKeystoreError)
+
+  try:
+
+    # We read all the file contents
+    var f: File
+    if not f.open(path, fmRead):
+      return err(OsError)
+    let fileContents = readAll(f)
+
+    # We iterate over each substring split by separator (which we expect to correspond to a single keystore json)
+    for keystore in fileContents.split(separator):
+
+      # We skip if read line is empty
+      if keystore.len == 0:
+        continue
+      # We skip all lines that don't seem to define a json
+      if not keystore.startsWith("{") or not keystore.endsWith("}"):
+        continue
+
+      try:
+        # We parse the json
+        data = json.parseJson(keystore)
+
+        # We check if parsed json contains the relevant keystore credentials fields and if these are set to the passed parameters 
+        # (note that "if" is lazy, so if one of the .contains() fails, the json fields contents will not be checked and no ResultDefect will be raised due to accessing unavailable fields)
+        if data.contains("application") and data.contains("appIdentifier") and data.contains("credentials") and data.contains("version") and
+           data["application"].getStr() == application and 
+           data["appIdentifier"].getStr() == appIdentifier and 
+           data["version"].getStr() == version:
```

Comment:
Can you create a proc which checks if `data` contains all of the elements in ["application", "appIdentifier", "credentials", "version"]?

```nim
proc hasAllKeys(data: Table[string, JsonNode], keys: seq[string]): bool =
  return all(data.contains(key) : key in keys)
```